File: 0x08-python-more_classes/4-rectangle.py
#!/usr/bin/python3
"""a class Rectangle that defines a rectangle by: (based on 0-rectangle.py)"""


class Rectangle:
    """ a class that defines the a rectangle
        properties: getter and setter methods for width and height
        methods: __init__ method -for creating a class
                area method - for calculating the area of the rectangle
                perimeter method -for calculating the perimeter of rectangle
                __str__ methods - that returns the string representation of
                Rectangle instance(s)
                _repr__ method - returns the formal strig representation of
                an object that can be evaluated
    """

    def __init__(self, width=0, height=0):
        self.height = height
        self.width = width

    def __str__(self):
        if self.width == 0 or self.height == 0:
            return ""
        result = ""
        for i in range(self.height):
            result += "#" * self.width + "\n"
        return result.rstrip()

    def __repr__(self):
        return f"Rectangle({self.width}, {self.height})"

    @property
    def width(self):
        return self.__width

    @width.setter
    def width(self, value):
        if not isinstance(value, int):
            raise TypeError("width must be an integer")
        if value < 0:
            raise ValueError("width must be >= 0")
        self.__width = value

    @property
    def height(self):
        return self.__height

    @height.setter
    def height(self, value):
        if not isinstance(value, int):
            raise TypeError("height must be an integer")
        if value < 0:
            raise ValueError("height must be >= 0")
        self.__height = value

    def area(self):
        return self.width * self.height

    def perimeter(self):
        if self.width == 0 or self.height == 0:
            return 0
        result = (self.width * 2) + (self.height * 2)
        return result

File: 0x13-javascript_objects_scopes_closures/4-rectangle.js
#!/usr/bin/node

/* Initializing  a class.
 * Adding methods rotate and double
 */

class Rectangle {
  constructor (w, h) {
    if (w > 0 && h > 0) {
      this.width = w;
      this.height = h;
    }
  }

  print () {
    let string = '';
    for (let i = 0; i < this.width; i++) {
      string += 'X';
    }
    for (let i = 0; i < this.height; i++) {
      console.log(string);
    }
  }

  rotate () {
    let tmp = '';
    tmp = this.width;
    this.width = this.height;
    this.height = tmp;
  }

  double () {
    this.width = this.width * 2;
    this.height = this.height * 2;
  }
}
module.exports = Rectangle;
